Port `cumsum` kernel to structured kernels.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61899 Port `cumprod` kernel to structured kernels.
* **#61842 Port `cumsum` kernel to structured kernels.**

Tracking issue: #55070

This PR also removes `at::_cumsum`, which was the "backend" for `at::cumsum`.

Differential Revision: [D29939152](https://our.internmc.facebook.com/intern/diff/D29939152)

ysiraichi committed:
Port `cumsum` kernel to structured kernels.

Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cumsum` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cumsum` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cumsum` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cumsum` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cumsum` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cumsum` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cumsum` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cumsum` kernel to structured kernels."


Tracking issue: #55070

This PR also removes `at::_cumsum`, which was the "backend" for `at::cumsum`.

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `cumsum` kernel to structured kernels."


Tracking issue: #55070

This PR also removes `at::_cumsum`, which was the "backend" for `at::cumsum`.

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29939152).

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -116,6 +116,31 @@ TORCH_META_FUNC(argmin)
+  if (result.defined()) {
+    out_dtype = dtype.value_or(result.scalar_type());
+    // This check is still here because the inline version of structured kernels
+    // does not do any checks on 'set_output'.
```
I'm not sure I understand. We do test that the passed dtype matches the out dtype; sample generated code:

```
    void set_output(int64_t output_idx, IntArrayRef sizes, IntArrayRef strides,
                    TensorOptions options, DimnameList names) override {
        
        const auto& out = outputs_[output_idx].get();
        TORCH_CHECK(options.dtype() == out.dtype(),
            "Expected out tensor to have dtype ", options.dtype(), ", but got ", out.dtype(), " instead");        TORCH_CHECK(options.device() == out.device(),
            "Expected out tensor to have device ", options.device(), ", but got ", out.device(), " instead");     
        bool resized = at::native::resize_output(outputs_[output_idx], sizes);

```

ysiraichi replied:
Ah. Replace `inline` => `inplace`. In the inplace version, we don't test it, do we?

ezyang replied:
Oh yeah, inplace still not testing; see also https://github.com/pytorch/pytorch/issues/57827

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -180,53 +205,21 @@ Tensor& logcumsumexp_out(const Tensor& self, int64_t dim, Tensor& result) {
+  } else {
+    dim = maybe_wrap_dim(dim, self.dim());
+    auto& result_mut = const_cast<Tensor&>(result);
+    cumsum_stub(self.device().type(), result_mut, self.to(result.scalar_type()), dim);
```
If possible, changing the type signature of `cumsum_stub` would be better, as these tend to be fairly localized changes.

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29939152).

ezyang committed:
Update on "Port `cumsum` kernel to structured kernels."


Tracking issue: #55070

This PR also removes `at::_cumsum`, which was the "backend" for `at::cumsum`.

Differential Revision: [D29939152](https://our.internmc.facebook.com/intern/diff/D29939152)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29939152).

ysiraichi committed:
Update on "Port `cumsum` kernel to structured kernels."


Tracking issue: #55070

This PR also removes `at::_cumsum`, which was the "backend" for `at::cumsum`.

Differential Revision: [D29939152](https://our.internmc.facebook.com/intern/diff/D29939152)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29939152).

ysiraichi commented:
The 2 CI failures do not seem to be related to this PR (not sure about the Facebook internal tests, though).
